### PR TITLE
Support aggregation pushdown in MySQL connector

### LINKED ISSUE #4138: Aggregation pushdown in MySQL connector
Follow-up to https://github.com/prestosql/presto/pull/3881

Relates to: https://github.com/prestosql/presto/issues/4139

### PR DESCRIPTION
**Pull request by yuokada**
Resolve #4138

### CODE REVIEW

**Comment by yuokada** [comment]
I'm not sure why the below incomplete query was executed. 🤔 


> 2020-07-12T15:11:03.8392009Z 2020-07-12T10:11:03.819-0500 SEVERE Error processing Split 20200712_151103_00103_3s5dq.1.0-0 io.prestosql.plugin.jdbc.JdbcSplit@b229a4c (start = 1466343.240782, wall = 7 ms, cpu = 0 ms, wait = 0 ms, calls = 1): JDBC_ERROR: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'double)) AS `_presto_generated_1` FROM `tpch`.`nation` GROUP BY `regionkey`' at line 1
2020-07-12T15:11:03.8475391Z io.prestosql.spi.PrestoException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'double)) AS `_presto_generated_1` FROM `tpch`.`nation` GROUP BY `regionkey`' at line 1



In my laptop, the AVG function is rewritten properly.

```sql
presto:classicmodels> EXPLAIN Select  AVG(amount_bigint) FROM x GROUP BY customernumber LIMIT 10;
                                                                                                 Query Plan
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 0 [SINGLE]
     Output layout: [_presto_generated_7]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     Output[_col0]
     │   Layout: [_presto_generated_7:double]
     │   Estimates: {rows: 10 (90B), cpu: ?, memory: 0B, network: 90B}
     │   _col0 := _presto_generated_7
     └─ Limit[10]
        │   Layout: [_presto_generated_7:double]
        │   Estimates: {rows: 10 (90B), cpu: ?, memory: 0B, network: 90B}
        └─ LocalExchange[SINGLE] ()
           │   Layout: [_presto_generated_7:double]
           │   Estimates: {rows: 10 (90B), cpu: ?, memory: 0B, network: 90B}
           └─ RemoteSource[1]
                  Layout: [_presto_generated_7:double]

 Fragment 1 [SOURCE]
     Output layout: [_presto_generated_7]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     LimitPartial[10]
     │   Layout: [_presto_generated_7:double]
     │   Estimates: {rows: 10 (90B), cpu: ?, memory: 0B, network: 0B}
     └─ TableScan[myold_mysql:classicmodels.x classicmodels.x columns=[avg(CAST(`amount_bigint` AS double)):_presto_generated_1:double:double] groupingSets=[[customernumber:integer:INT]], grouped = false]
            Layout: [_presto_generated_7:double]
            Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}
            _presto_generated_7 := avg(CAST(`amount_bigint` AS double)):_presto_generated_1:double:double


(1 row)
```

**Comment by yuokada** [comment]
I roughly implemented. Could you please take a look once?

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySQLClient.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    private static final JdbcColumnHandle DOUBLE_COLUMN =
+            JdbcColumnHandle.builder()
+                    .setColumnName("c_double")
+                    .setColumnType(DOUBLE)
+                    .setJdbcTypeHandle(new JdbcTypeHandle(Types.DOUBLE, Optional.of("double"), 0, 0, Optional.empty()))
+                    .build();
+
+    private static final JdbcClient JDBC_CLIENT = new MySqlClient(
+            new BaseJdbcConfig(),
+//            new MySqlConfig(),
```
[the remark this answers is not in the document]
Comment (yuokada) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySQLClient.java]: Fixed

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java [hunk 2]
```diff
@@ -225,6 +225,34 @@ public void testInsertIntoNotNullColumn()
...
+        // TODO support aggregation pushdown with GROUPING SETS
+        // TODO support aggregation over expressions
+
+        assertPushedDown("SELECT count(*) FROM nation");
+        assertPushedDown("SELECT count(nationkey) FROM nation");
+        assertPushedDown("SELECT regionkey, min(nationkey) FROM nation GROUP BY regionkey");
+        assertPushedDown("SELECT regionkey, max(nationkey) FROM nation GROUP BY regionkey");
+        assertPushedDown("SELECT regionkey, sum(nationkey) FROM nation GROUP BY regionkey");
+        assertPushedDown("SELECT regionkey, avg(regionkey) FROM nation GROUP BY regionkey");
+        assertPushedDown(
```
[the remark this answers is not in the document]
Comment (yuokada) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java:241]: I will rebase.

[the remark this answers is not in the document]
Comment (yuokada) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java]: Rebased & Renamed.

File: presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import static io.prestosql.plugin.jdbc.expression.AggregateFunctionPatterns.expressionType;
+import static io.prestosql.plugin.jdbc.expression.AggregateFunctionPatterns.functionName;
+import static io.prestosql.plugin.jdbc.expression.AggregateFunctionPatterns.singleInput;
+import static io.prestosql.plugin.jdbc.expression.AggregateFunctionPatterns.variable;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.DoubleType.DOUBLE;
+import static io.prestosql.spi.type.IntegerType.INTEGER;
+import static java.lang.String.format;
+
+public class ImplementAvgBigint
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java:42]: Let's not block on this. It's not a big problem to adapt later.

File: presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    private static final Capture<Variable> INPUT = newCapture();
+
+    @Override
+    public Pattern<AggregateFunction> getPattern()
+    {
+        return basicAggregation()
+                .with(functionName().equalTo("avg"))
+                .with(singleInput().matching(
+                        variable()
+                                .with(expressionType().matching(type -> type == BIGINT || type == INTEGER))
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java:54]: good catch

File: presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    @Override
+    public Optional<JdbcExpression> rewrite(AggregateFunction aggregateFunction, Captures captures, RewriteContext context)
+    {
+        Variable input = captures.get(INPUT);
+        JdbcColumnHandle columnHandle = (JdbcColumnHandle) context.getAssignments().get(input.getName());
+        verifyNotNull(columnHandle, "Unbound variable: %s", input);
+        verify(aggregateFunction.getOutputType() == DOUBLE);
+
+        return Optional.of(new JdbcExpression(
+                format("avg(CAST(%s AS double))", columnHandle.toSqlExpression(context.getIdentifierQuote())),
```
Comment (ebyhr) [inline presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java:67]: Could you please fix the CI failure?
```
2020-07-24T00:55:02.4084660Z [ERROR]   TestMySqlDistributedQueries>AbstractTestQueries.testExistsSubquery:1025->AbstractTestQueryFramework.assertQuery:120->AbstractTestQueryFramework.assertQuery:125 Execution of 'actual' query failed: SELECT o1.orderkey, COUNT(*) FROM orders o1  JOIN (SELECT * FROM orders LIMIT 10) o2 ON (SELECT avg(orderkey) FROM orders) > 3 GROUP BY o1.orderkey ORDER BY o1.orderkey LIMIT 5
2020-07-24T00:55:02.4086032Z [ERROR]   TestMySqlDistributedQueries>AbstractTestQueries.testScalarSubquery:907->AbstractTestQueryFramework.assertQuery:120->AbstractTestQueryFramework.assertQuery:125 Execution of 'actual' query failed: SELECT DISTINCT orderkey FROM lineitem WHERE orderkey BETWEEN   (SELECT avg(orderkey) FROM orders) - 10    AND   (SELECT avg(orderkey) FROM orders) + 10
2020-07-24T00:55:02.4086880Z [ERROR]   TestMySqlIntegrationSmokeTest>AbstractTestIntegrationSmokeTest.testAggregation:104->AbstractTestQueryFramework.assertQuery:130 Execution of 'actual' query failed: SELECT avg(nationkey) FROM nation
2020-07-24T00:55:02.4087788Z [ERROR]   TestMySqlIntegrationSmokeTest.testAggregationPushdown:242->AbstractTestIntegrationSmokeTest.assertAggregationPushedDown:452->AbstractTestQueryFramework.computeActual:105->AbstractTestQueryFramework.computeActual:110 » Runtime
```

```
mysql> SELECT cast(1 AS DOUBLE);
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'DOUBLE)' at line 1
```

Comment (yuokada) [reply presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java:67]: Yeah, I'm investigating the cause of it.

Comment (yuokada) [reply presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java:67]: @ebyhr Just a question. what version of mysql did you use to reproduce it?
I can't reproduce it in mysql 8.0.
 
```
mysql> SELECT cast(1 AS DOUBLE);
+-------------------+
| cast(1 AS DOUBLE) |
+-------------------+
|                 1 |
+-------------------+
1 row in set (0.00 sec)

mysql> select version();
+-----------+
| version() |
+-----------+
| 8.0.19    |
+-----------+
1 row in set (0.00 sec)
```

Comment (ebyhr) [reply presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java:67]: We are using version 8.0.12.
https://github.com/prestosql/presto/blob/337dd30d26a0ace0b5b2bd99ccedcc0a556f35de/presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestingMySqlServer.java#L30

Comment (yuokada) [reply presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java:67]: I see. And you are right.
> DOUBLE
Produces a DOUBLE result. Added in MySQL 8.0.17.

https://dev.mysql.com/doc/refman/8.0/en/cast-functions.html#function_cast

File: presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuokada) [reply presto-mysql/src/main/java/io/prestosql/plugin/mysql/ImplementAvgBigint.java]: @alexjo2144 I confirmed aggregate pushdown isn't triggered to INTEGER in mysql and postgresql.
So I revised.

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java [hunk 6]
```diff
@@ -226,6 +226,33 @@ public void testInsertIntoNotNullColumn()
...
+
+        assertThat(query("SELECT count(*) FROM nation")).isCorrectlyPushedDown();
+        assertThat(query("SELECT count(nationkey) FROM nation")).isCorrectlyPushedDown();
+        assertThat(query("SELECT count(1) FROM nation")).isCorrectlyPushedDown();
+        assertThat(query("SELECT count() FROM nation")).isCorrectlyPushedDown();
+        assertThat(query("SELECT regionkey, min(nationkey) FROM nation GROUP BY regionkey")).isCorrectlyPushedDown();
+        assertThat(query("SELECT regionkey, max(nationkey) FROM nation GROUP BY regionkey")).isCorrectlyPushedDown();
+        assertThat(query("SELECT regionkey, sum(nationkey) FROM nation GROUP BY regionkey")).isCorrectlyPushedDown();
+        assertThat(query("SELECT regionkey, avg(nationkey) FROM nation GROUP BY regionkey")).isCorrectlyPushedDown();
+
```
Comment (findepi) [inline presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java:244]: Add

```
assertThat(query("SELECT regionkey, sum(nationkey) FROM nation WHERE regionkey < 4 GROUP BY regionkey")).isCorrectlyPushedDown();
        assertThat(query("SELECT regionkey, sum(nationkey) FROM nation WHERE regionkey < 4 AND name > 'AAA' GROUP BY regionkey")).isNotFullyPushedDown(FilterNode.class);
```

see https://github.com/prestosql/presto/blob/1e69a5bfed50c7b41febcd62b79f598c52bf67d6/presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java#L352-L353

**Review by findepi — APPROVED** [review]
@yuokada overall LGTM, thanks!
one minor comment

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java (hunk 6 above)
Comment (yuokada) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java:244]: Thanks. I added those lines.

**Comment by mosabua** [comment]
This needs relevant documentation added similar to the postgresql one. https://prestosql.io/docs/current/connector/postgresql.html#pushdown

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySQLClient.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+import static io.prestosql.metadata.MetadataManager.createTestMetadataManager;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.BooleanType.BOOLEAN;
+import static io.prestosql.spi.type.DoubleType.DOUBLE;
+import static io.prestosql.testing.TestingConnectorSession.SESSION;
+import static io.prestosql.testing.assertions.Assert.assertEquals;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.assertTrue;
+
+public class TestMySQLClient
```
Comment (findepi) [inline presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySQLClient.java:44]: nit: `TestMySqlClient`

**Review by findepi — COMMENTED** [review]
sorry, one more thing i overlooked previosuly.

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySQLClient.java (hunk 7 above)
Comment (yuokada) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySQLClient.java]: Renamed.

**Comment by yuokada** [comment]
Hmm... tests for presto-postgresql is failing.
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for presto-mysql 341-SNAPSHOT:
[INFO] 
[INFO] presto-mysql ....................................... SUCCESS [07:26 min]
[INFO] presto-postgresql .................................. FAILURE [06:01 min]
[INFO] presto-sqlserver ................................... SUCCESS [10:26 min]
[INFO] ------------------------------------------------------------------------
```
link: https://github.com/prestosql/presto/runs/1028167123

File: presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java (hunk 6 above)
Comment (tooptoop4) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java:244]: @findepi why isNotFullyPushedDown is used for "SELECT regionkey, sum(nationkey) FROM nation WHERE regionkey < 4 AND name > 'AAA' GROUP BY regionkey" ?

Comment (findepi) [reply presto-mysql/src/test/java/io/prestosql/plugin/mysql/TestMySqlIntegrationSmokeTest.java:244]: @tooptoop4 https://github.com/prestosql/presto/blob/22717c61419c851b1c3814b60a3453b22d4a45ef/presto-mysql/src/main/java/io/prestosql/plugin/mysql/MySqlClient.java#L237-L238

**Pull request merged**